aspnetcore/web-api/action-return-types/samples/3x/WebApiSample.DataAccess/Repositories/ProductsRepository.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using WebApiSample.DataAccess.Models;
+
+namespace WebApiSample.DataAccess.Repositories
+{
+    public class ProductsRepository
+    {
+        private readonly ProductContext _context;
+
+        public ProductsRepository(ProductContext context)
+        {
+            _context = context;
+
+            if (_context.Products.Count() == 0)
+            {
+                _context.Products.AddRange(
+                    new Product
+                    {
+                        Name = "Learning ASP.NET Core",
+                        Description = "A best-selling book covering the fundamentals of ASP.NET Core"
+                    },
+                    new Product
+                    {
+                        Name = "Learning EF Core",
+                        Description = "A best-selling book covering the fundamentals of Entity Framework Core"
+                    },
+                    new Product
+                    {
+                        Name = "Learning .NET Standard",
+                        Description = "A best-selling book covering the fundamentals of .NET Standard"
+                    },
+                    new Product
+                    {
+                        Name = "Learning .NET Core",
+                        Description = "A best-selling book covering the fundamentals of .NET Core"
+                    },
+                    new Product
+                    {
+                        Name = "Learning C#",
+                        Description = "A best-selling book covering the fundamentals of C#"
+                    });
+                _context.SaveChanges();
+            }
+        }
+
+        public List<Product> GetProducts() =>
+            _context.Products.OrderBy(p => p.Name).ToList();
+
+        public IEnumerable<Product> GetProductsByPage(
```

Comment:
I would expect that this method returns a type that implements **both** IAsyncEnumerable<> and IEnumerable<>.

You're having trouble demonstrating the problem because we fixed the problem 😆 .

If you want a good sample of this, you have to put a foreach loop in both this and `GetProductsByPage` that does something in the loop.

Reply:
> I would expect that this method returns a type that implements both IAsyncEnumerable<> and IEnumerable<>.

@rynowak Do you have an example of such a collection type? Or would this be a custom collection type I create that implements both interfaces?

Reply:
@rynowak, can you clarify your expectation?  I expect it'll be super rare for a type to implement both interfaces.  The design of having an `IEnumerable<T> Foo` and `IAsyncEnumerable<T> FooAsync` is much more in line with what I'd expect to see.

Reply:
https://github.com/aspnet/EntityFrameworkCore/blob/24b9aa1d2e14fe2e737255ede9b2a7a623fcf2af/src/EFCore/DbSet.cs

**LOTS** of types in EF.

We did this feature to solve a very specific problem.

```C#
public IEnumerable<Project> GetOnSaleProducts()
{
    return db.Products.Where(p => p.IsOnSale);
}
```

In previous versions of MVC - this would **SYNCHRONOUSLY** enumerate the results of that query while serializing the data.

It's totally non-obvious to a user that this is doing blocking waits on the database. To write this code correctly in 2.2, you would write:

```C#
public IEnumerable<Project> GetOnSaleProducts()
{
    return db.Products.Where(p => p.IsOnSale).ToListAsync();
}
```

---

In 3.0 you can write the naive code - and we will buffer the results before handing it off to the serializer:

```C#
public IEnumerable<Project> GetOnSaleProducts()
{
    return db.Products.Where(p => p.IsOnSale);
}
```

This works because:
- MVC looks at the concrete type of the value returned by your action, and buffers it if it's `IAsyncEnumerable<>`
- The **declared** return types of EF's query stuff are `IQueryable<>` and related types, but the implementations of `IQueryable<>` all also implement `IAsyncEnumerable<>`.

This kind of design approach is necessary because `IAsyncEnumerable<>` was adding to existing APIs.

---

In 5.0 when S.T.J supports `IAsyncEnumerable<>` we'll remove the buffering and users can have end-to-end streaming.

Reply:
@scottaddie  - if you want a good example of this:

```c#
public IEnumerable<Product> GetOnSaleProducts()
{
    var products = repository.GetProducts();
    foreach (var product in products)
    {
        if (product.IsOnSale)
        {
             yield return product;
        }
    }
}

public async IAsyncEnumerable<Product> GetOnSaleProducts()
{
    var products = repository.GetProductsAsync();
    await foreach (var product in products)
    {
        if (product.IsOnSale)
        {
             yield return product;
        }
    }
}
```

Reply:
When you add some user code - the example becomes more useful

Reply:
> if you want a good example of this

Exactly.  This example doesn't have one method that returns something that implements both `IEnumerable<T>` and `IAsyncEnumerable<T>`; it has two methods, one returning `IEnumerable<T>` and one returning `IAsyncEnumerable<T>`.  Maybe I misunderstood your comment "I would expect that this method returns a type that implements both IAsyncEnumerable<> and IEnumerable<>."?

Reply:
I think we both misunderstood that comment, @stephentoub. Shout out to @rynowak for all of the help. I learned a lot of useful stuff through the work on this PR. I'm confident the community will benefit from this collaboration.

Reply:
Perhaps my statements were a little confusing. 😁  

This feature helps two cases:
- EF's return types that implement both APIs
- User-code that returns IAsyncEnumerable

I'm only separating these two because in the former case, it's totally non-obvious to a user that they were working with IAsyncEnumerable because it's not part of the declared type (IQueryable).